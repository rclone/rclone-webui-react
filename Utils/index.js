import checkPropTypes from 'check-prop-types';
import {applyMiddleware, createStore} from 'redux';
import rootReducer from './../src/reducers';
import {middleware} from '../src/store';

export const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
};

export const checkProps = (component, expectedProps) => {
    const propsErr = checkPropTypes(component.propTypes, expectedProps, 'prop', component.name);
    return propsErr;
};

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};