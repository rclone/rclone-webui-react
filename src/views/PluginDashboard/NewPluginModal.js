import React, {useState} from 'react';
import {
    Button,
    Col,
    Form,
    FormFeedback,
    FormGroup,
    Input,
    Label,
    Modal,
    ModalBody,
    ModalFooter,
    ModalHeader
} from 'reactstrap';
import * as PropTypes from "prop-types"

/**
 * NewPluginModal shows a button for opening a modal for new mount and then executes okHandle when positive
 * button is clicked
 * @param props
 * @returns {*}
 * @constructor
 */
const NewPluginModal = (props) => {
    const {
        buttonLabel,
        className,
        okHandle
    } = props;

    const [modal, setModal] = useState(false);

    const [pluginDownloadURL, setPluginDownloadURL] = useState("");

    const [pluginType, setPluginType] = useState("");

    const toggle = () => setModal(!modal);

    const handleCreateMount = () => {
        if (!okHandle) {
            throw new Error("Ok handle is null");
        }
        okHandle(pluginDownloadURL, pluginType);
    }

    const isCreateDisabled = () => {
        return !pluginDownloadURL || pluginDownloadURL === "";
    }

    const submitForm = (e) => {
        e.preventDefault();
        handleCreateMount();
    }

    return (
        <div data-test="newMountModalComponent">
            <Button color="primary" onClick={toggle}>{buttonLabel}</Button>
            <Form onSubmit={submitForm}>
                <Modal isOpen={modal} toggle={toggle} className={className}>
                    <ModalHeader toggle={toggle}>New Mount</ModalHeader>
                    <ModalBody>

                        <FormGroup row>
                            <Label for={"pluginURL"} sm={5}>Plugin URL</Label>
                            <Col sm={7}>
                                <Input type={"text"} value={pluginDownloadURL}
                                       name={"pluginURL"}
                                       id={"pluginURL"} onChange={e => setPluginDownloadURL(e.target.value)}
                                       required={true}
                                >
                                </Input>
                                <FormFeedback/>
                            </Col>
                        </FormGroup>

                        <FormGroup row>
                            <Label for={"pluginType"} sm={5}>Plugin Type</Label>
                            <Col sm={7}>
                                <Input type={"text"} value={pluginDownloadURL}
                                       name={"pluginType"}
                                       id={"pluginType"} onChange={e => setPluginType(e.target.value)}
                                       required={true}>
                                </Input>
                                <FormFeedback/>
                            </Col>
                        </FormGroup>
                    </ModalBody>
                    <ModalFooter>
                        <Button data-test="ok-button" color="primary" onClick={handleCreateMount}
                                disabled={isCreateDisabled()}>Add</Button>{' '}
                        <Button data-test="cancel-button" color="secondary" onClick={toggle}>Cancel</Button>
                    </ModalFooter>
                </Modal>
            </Form>
        </div>
    );
}

NewPluginModal.propTypes = {
    /**
     * Text for open modal button
     */
    buttonLabel: PropTypes.string,
    /**
     * Class for open modal button
     */
    buttonClass: PropTypes.string,
    /**
     * Function to be called when ok button is clicked.
     */
    okHandle: PropTypes.func.isRequired,
}

export default NewPluginModal;