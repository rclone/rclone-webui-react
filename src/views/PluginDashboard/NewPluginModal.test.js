import React from "react";
import {shallow} from "enzyme";
import {findByTestAttr, testStore} from "../../../Utils";
import toJson from "enzyme-to-json";
import NewPluginModal from "./NewPluginModal";


const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);

    const component = shallow(<NewPluginModal {...props} store={store}/>);
    return component;
};


describe('New Plugin Modal', function () {
    describe('renders', function () {
        const myMock = jest.fn();
        let wrapper;
        beforeEach(() => {
            const initialState = {};

            const props = {
                okHandle: myMock
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            const component = findByTestAttr(wrapper, "newMountModalComponent");
            expect(component).toHaveLength(1);
        });

        it('should have ok and cancel buttons', function () {
            const okButton = findByTestAttr(wrapper, "ok-button")
            expect(okButton).toHaveLength(1);

            const cancelButton = findByTestAttr(wrapper, "cancel-button");
            expect(cancelButton).toHaveLength(1);
        })


        it('okHandle should have been called once', function () {
            const okButton = findByTestAttr(wrapper, "ok-button")
            okButton.simulate('click');

            expect(myMock.mock.calls).toHaveLength(1);
        })

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot();
        });

    });
});
