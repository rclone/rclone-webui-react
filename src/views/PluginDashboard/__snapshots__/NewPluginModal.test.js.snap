// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Plugin Modal renders should match snapshot 1`] = `
<div
  data-test="newMountModalComponent"
>
  <Button
    color="primary"
    onClick={[Function]}
    tag="button"
  />
  <Form
    onSubmit={[Function]}
    tag="form"
  >
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      container="body"
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      returnFocusAfterClose={true}
      role="dialog"
      scrollable={false}
      toggle={[Function]}
      unmountOnClose={true}
      zIndex={1050}
    >
      <ModalHeader
        charCode={215}
        closeAriaLabel="Close"
        tag="h5"
        toggle={[Function]}
        wrapTag="div"
      >
        New Mount
      </ModalHeader>
      <ModalBody
        tag="div"
      >
        <FormGroup
          row={true}
          tag="div"
        >
          <Label
            for="pluginURL"
            sm={5}
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Plugin URL
          </Label>
          <Col
            sm={7}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Input
              id="pluginURL"
              name="pluginURL"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
            <FormFeedback
              tag="div"
            />
          </Col>
        </FormGroup>
        <FormGroup
          row={true}
          tag="div"
        >
          <Label
            for="pluginType"
            sm={5}
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Plugin Type
          </Label>
          <Col
            sm={7}
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Input
              id="pluginType"
              name="pluginType"
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
            <FormFeedback
              tag="div"
            />
          </Col>
        </FormGroup>
      </ModalBody>
      <ModalFooter
        tag="div"
      >
        <Button
          color="primary"
          data-test="ok-button"
          disabled={true}
          onClick={[Function]}
          tag="button"
        >
          Add
        </Button>
         
        <Button
          color="secondary"
          data-test="cancel-button"
          onClick={[Function]}
          tag="button"
        >
          Cancel
        </Button>
      </ModalFooter>
    </Modal>
  </Form>
</div>
`;
